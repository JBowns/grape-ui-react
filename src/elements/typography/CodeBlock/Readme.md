When showing example code, `<CodeBlock>` is the component you should use.  It is built on top of the powerful [react-syntax-highlighter](https://github.com/conorhastings/react-syntax-highlighter) and uses the same utilities that can be found on all typography elements in grape-ui.  Just provide your code as a string in the `codeString` prop, and you're ready to go.

### Examples

```jsx in Markdown
const exampleCode01 = `<CodeBlock codeString={exampleCode01} />`;
const exampleCode02 = `
import React from 'react';
import { Paragraph } from 'grape-ui-react';
const App = () => {
  return (
    <div className="App">
      <Paragraph>
        grape-ui is Groovy
      </Paragraph>
    </div>
  )
};
export default App;

`;
const exampleCode03 = "npm i grape-ui-react";
const exampleCode04 = `

![grape-ui logo](http://www.grapeui.com/assets/images/grape-ui-header-logo.svg)

## Getting Started

These instructions will allow you to use the extensible grape-ui component library in any React project.

### Prerequisites

* [ReactJS](https://reactjs.org/)
* [Styled Components](https://www.styled-components.com/)

### Installing

Add grape-ui as a dependency to your project

\`\`\`bash
npm install grape-ui-react
\`\`\`

### Usage

Import grape-ui components into your project

\`\`\`jsx static
import { Paragraph } from 'grape-ui-react';
\`\`\`

Use the components in your app

\`\`\`jsx static
render() {
  return (
    <div className="App">
      ...
      <Paragraph>grape-ui is Groovy</Paragraph>
      ...
    </div>
  );
};
\`\`\`

And that's it!

`;

<div>
  <CodeBlock
    codeString={exampleCode01}
    display="inline-block"
  />
  <CodeBlock codeString={exampleCode02} />
  <CodeBlock codeString={exampleCode03} language="bash" />
  <CodeBlock codeString={exampleCode04} language="markdown" />
</div>
```

For additional tools and info, please refer to [react-syntax-highlighter's docs](https://github.com/conorhastings/react-syntax-highlighter).
