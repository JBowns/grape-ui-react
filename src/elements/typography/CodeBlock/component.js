import React from 'react';
import PropTypes from 'prop-types';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { removeSomeProps } from 'src/utils/componentHelpers';
import { grapeUICustomStyle, codeBlockPropTypes } from './utils';

export const CodeBlockComponent = ({
  codeString,
  language,
  style,
  ...props
}) => (
  <SyntaxHighlighter
    language={language}
    style={style}
    {...removeSomeProps(props, Object.keys(codeBlockPropTypes))}
  >
    {codeString}
  </SyntaxHighlighter>
);

CodeBlockComponent.propTypes = {
  codeString: PropTypes.string,
  language: PropTypes.string,
  style: PropTypes.any,
};

CodeBlockComponent.defaultProps = {
  codeString: '',
  language: 'javascript',
  style: grapeUICustomStyle,
};
