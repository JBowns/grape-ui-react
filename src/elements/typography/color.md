There are many colors available to be used in grape-ui. Just add the `color="colorName"` attribute to any text element.

| colorProp | Description |
| - | - |
| `colorName` or `colorName.base` | Returns the defined base color |
| `colorName.light` | Returns the defined light variant |
| `colorName.dark` | Returns the defined dark variant |
| `else` | Returns provided value |

### List of Colors Available

```jsx in Markdown
import { getGlobalStyles } from 'src/global-styles'; // ... from 'grape-ui-react'
import {
  List,
  ListItem,
  Text,
} from './';

renderColorListItems = () => {
  const { colors } = getGlobalStyles();
  const colorKeys = Object.keys(colors).filter(
    prop => prop !== 'default'
  ); // [ prop names of colors... ]
  return colorKeys.map(cKey =>
    <ListItem
      color={cKey}
      key={cKey}
      style={{
        textTransform: 'capitalize'
      }}
    >
      {`${cKey} (`}
      <Text
        color={`${cKey}.light`}
      >
        Light
      </Text>
      /
      <Text
        color={`${cKey}.dark`}
      >
        Dark
      </Text>
      )
    </ListItem>
  );
}

<List>
  {this.renderColorListItems()}
</List>
```

### Style Overrides

To overwrite the colors, or introduce any new colors, simply call `colors` followed by the name of the color, and defining its `base`, `dark`, and `light` variants.

```jsx in Markdown
import { ThemeProvider } from 'styled-components';
import { Paragraph, Text } from './index';

const customBrandPrimaryTheme = {
  colors: {
    brandPrimary: {
      base: 'hsl(143.3, 84.6%, 28%)',
      dark: 'hsl(143.3, 84.6%, 8%)',
      light: 'hsl(143.3, 84.6%, 48%)',
    },
  },
};

<div>
  <Paragraph
    color="brandPrimary"
  >
    This paragraph is set to the default brand primary color,
    <Text color="brandPrimary.dark">
      {' dark variant'}
    </Text>
    <Text color="brandPrimary.light">
      , and the light variant
    </Text>
    .
  </Paragraph>
  <ThemeProvider
    theme={customBrandPrimaryTheme}
  >
    <Paragraph
      color="brandPrimary"
    >
      This paragraph is set to a custom brand primary color,
      <Text color="brandPrimary.dark">
        {' dark variant'}
      </Text>
      <Text color="brandPrimary.light">
        , and the light variant
      </Text>
      .
    </Paragraph>
  </ThemeProvider>
</div>
```
