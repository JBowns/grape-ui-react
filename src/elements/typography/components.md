## [`<Header>`](#/Typography/Components/Header)

Learn about our header title components, including display headers and style overrides for it.
***

## [`<Link>`](#/Typography/Components/Link)

Learn about link components, including how it can be used with [react-router](https://reacttraining.com/react-router/).
***

## [`<List>`](#/Typography/Components/List)

Learn about list and listitem components, including how it can be used as an ordered list and style variations.
***

## [`<Paragraph>`](#/Typography/Components/Paragraph)

Learn about paragraph components, including lead paragraphs.
***

## [`<Text>`](#/Typography/Components/Text)

Learn about inline text components, including inline styling.
***
