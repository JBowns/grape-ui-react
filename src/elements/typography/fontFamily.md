grape-ui has five different font families that it uses. Simply call to them by entering `fontFamily="fontFamilyName"` as an attribute to any text component. By default, we use `base` as the system-wide font family.

### List of font families available

```jsx in Markdown
import { getGlobalStyles } from 'src/global-styles'; // ... from 'grape-ui-react'
import { List, ListItem, Text } from './index';

renderFontFamilyListItems = () => {
  const { fontFamily } = getGlobalStyles();
  const fontFamilyKeys = Object.keys(fontFamily).filter(prop => prop !== 'default'); // [ prop names of font families... ]
  return fontFamilyKeys.map(cKey =>
    <ListItem fontFamily={cKey} key={cKey} style={{ textTransform: 'capitalize' }}>{cKey}</ListItem>);
}

<List>{this.renderFontFamilyListItems()}</List>
```

### Style Overrides

To overwrite the font family definitions, or introduce any new families, simply call `fonts` followed by the name of the font stack, and defining its font families.

```jsx in Markdown
import { ThemeProvider } from 'styled-components';
import { Paragraph, Text } from './index';

const customBaseFontFamilyTheme = {
  fonts: {
    base: 'Papyrus,fantasy',
  },
};

<div>
  <Paragraph>This paragraph is using the default base font family.</Paragraph>
  <ThemeProvider theme={customBaseFontFamilyTheme}>
    <Paragraph>This paragraph is using a custom base font family.</Paragraph>
  </ThemeProvider>
</div>
```
