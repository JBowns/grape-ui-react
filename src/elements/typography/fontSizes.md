To update font sizes, you can try adding `lg` for large text and `sm` for small text.

```jsx in Markdown
import { Header, Paragraph } from './index';

<div>
  <Paragraph lg>This paragraph is large.</Paragraph>
  <Paragraph>This paragraph is normal.</Paragraph>
  <Paragraph sm>This paragraph is small.</Paragraph>
  <Header.h5 lg>This Header.h5 is large.</Header.h5>
  <Header.h5>This Header.h5 is normal.</Header.h5>
  <Header.h5 sm>This Header.h5 is small.</Header.h5>
  <Header.h5 displayHeader lg>This Display Header.h5 is large.</Header.h5>
  <Header.h5 displayHeader>This Display Header.h5 is normal.</Header.h5>
  <Header.h5 displayHeader sm>This Display Header.h5 is small.</Header.h5>
</div>
```