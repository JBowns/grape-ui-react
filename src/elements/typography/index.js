export * from './Header';
export * from './List';
export * from './ListItem';
export * from './Paragraph';
export * from './Text';
export * from './Link';
