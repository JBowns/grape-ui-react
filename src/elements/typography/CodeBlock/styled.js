import styled from 'styled-components';
import {
  background,
  border,
  fontWeight,
  layout,
  space,
} from 'styled-system';
import {
  colorCore,
  ellipsisCore,
  fontFamilyCore,
  fontSizeCore,
  fontStyleCore,
  letterSpacingCore,
  lineHeightCore,
  textAlignCore,
  textDecorationCore,
} from 'src/utils/styledHelpers';
import { CodeBlockComponent } from './component';
import { codeBlockDefaultProps, codeBlockPropTypes } from './utils';

const CodeBlock = styled(CodeBlockComponent)`
  ${background}
  ${border}
  ${colorCore}
  ${ellipsisCore}
  ${fontSizeCore}
  ${fontStyleCore}
  ${fontWeight}
  ${layout}
  ${letterSpacingCore}
  ${lineHeightCore}
  ${space}
  ${textAlignCore}
  ${textDecorationCore}
  > code {
    ${fontFamilyCore}
  }
`;

CodeBlock.propTypes = {
  ...codeBlockPropTypes,
};

CodeBlock.defaultProps = {
  ...codeBlockDefaultProps,
};

/** @component */
export { CodeBlock };
