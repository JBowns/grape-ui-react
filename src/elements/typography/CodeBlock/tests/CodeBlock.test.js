import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { CodeBlock } from 'src/elements/typography/CodeBlock';

const emptyTheme = {};

const assertReactElement = element => {
  const component = renderer.create(element);
  return component.toJSON();
};

describe('Box Component base', () => {
  it('should return an empty Toolbar with base styling', () => {
    const element = (
      <ThemeProvider theme={emptyTheme}>
        <CodeBlock codeString={'<CodeBlock codeString="" />'} />
      </ThemeProvider>
    );
    expect(assertReactElement(element)).toMatchSnapshot();
  });
});
