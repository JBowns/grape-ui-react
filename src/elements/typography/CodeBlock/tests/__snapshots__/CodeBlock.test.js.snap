// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box Component base should return an empty Toolbar with base styling 1`] = `
.c0 {
  background: whitesmoke;
  border: 1px solid #eee;
  border-radius: 4px;
  background-color: transparent;
  color: hsl(0,0%,25%);
  font-size: 0.8rem;
  font-weight: inherit;
  display: block;
  overflow-x: auto;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  line-height: 1.5;
  margin-top: 4px;
  margin-bottom: 4px;
  padding: 4px;
  text-align: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
}

.c0 > code {
  font-family: Consolas,"Andale Mono WT","Andale Mono","Lucida Console","Lucida Sans Typewriter","DejaVu Sans Mono","Bitstream Vera Sans Mono","Liberation Mono","Nimbus Mono L",Monaco,"Courier New",Courier,monospace;
}

@media screen and (min-width:40em) {
  .c0 {
    padding: 8px;
  }
}

@media screen and (min-width:52em) {
  .c0 {
    margin-top: 8px;
    margin-bottom: 8px;
    padding: 16px;
  }
}

<pre
  className="c0"
  style={
    Object {
      "background": "hsl(0, 0%, 98%)",
    }
  }
>
  <code>
    &lt;CodeBlock codeString=
    <span
      style={
        Object {
          "color": "hsl(174.4, 100%, 29.4%)",
        }
      }
    >
      ""
    </span>
     /&gt;
  </code>
</pre>
`;
